crates/sui-core/src/validator_client_monitor/stats.rs

```diff
@@ -127,149 +128,73 @@ impl ClientObservedStats {
-    fn calculate_client_score(
-        &self,
-        stats: &ValidatorClientStats,
-        max_latencies: &HashMap<OperationType, f64>,
-        tx_type: TxType,
-    ) -> f64 {
-        let mut latency_score = 0.0;
-        let mut total_weight = 0.0;
+    /// Returns latency in seconds, with reliability penalty applied as a multiplier.
+    fn calculate_client_latency(&self, validator: &AuthorityName, tx_type: TxType) -> f64 {
+        let mut final_latency = MAX_LATENCY;
 
-        for op in OperationType::iter() {
-            let latency_weight = match op {
-                OperationType::Submit => self.config.score_weights.submit_latency_weight,
-                OperationType::Effects => self.config.score_weights.effects_latency_weight,
-                OperationType::HealthCheck => self.config.score_weights.health_check_latency_weight,
-                OperationType::FastPath => self.config.score_weights.fast_path_latency_weight,
-                OperationType::Consensus => self.config.score_weights.consensus_latency_weight,
-            };
+        let Some(stats) = self.validator_stats.get(validator) else {
+            return final_latency;
+        };
 
-            if tx_type == TxType::SingleWriter && op == OperationType::Consensus {
-                continue;
+        if let Some(exclusion_time) = stats.exclusion_time {
+            if exclusion_time.elapsed() < self.config.failure_cooldown {
+                return final_latency;
             }
+        }
 
-            if tx_type == TxType::SharedObject && op == OperationType::FastPath {
-                continue;
-            }
+        let operation = match tx_type {
+            TxType::SharedObject => OperationType::Consensus,
+            TxType::SingleWriter => OperationType::FastPath,
+        };
 
-            // Skip if max latency is missing for this operation
-            let Some(max_latency) = max_latencies.get(&op) else {
-                continue;
+        // Get the latency for the relevant operation
+        if let Some(latency) = stats.average_latencies.get(&operation) {
+            let base_latency = latency.get();
+            let reliability = stats.reliability.get();
+            let reliability_weight = self.config.reliability_weight;
+
+            // Apply reliability penalty based on the configured weight
+            // reliability_weight = 0.0: no reliability consideration, use base latency
+            // reliability_weight = 1.0: full reliability penalty, divide by reliability
+            // reliability_weight in (0.0, 1.0): interpolate between base latency and reliability-penalized latency
+            let penalty_factor = if reliability > 0.0 && reliability_weight > 0.0 {
+                // Calculate the penalty factor: 1.0 (no penalty) to 1.0/reliability (full penalty)
+                let max_penalty = 1.0 / reliability;
+                1.0 + reliability_weight * (max_penalty - 1.0)
+            } else {
+                1.0
```

Comment:
Not sure I agree with this. When `reliability` is 0, the penalty should be max. But in this case the penalty factor is the minimum.